include/seastar/util/sampler.hh

```diff
@@ -145,14 +152,12 @@ private:
         int64_t next = static_cast<int64_t>(dist(random_gen));
         // We approximate the geometric distribution using an exponential
         // distribution.
-        // We need to add 1 because that gives us the number of failures before
-        // the next success, while our interval includes the next success.
-        return next + 1;
+        return next;
```

Comment:
The important part is that `maybe_sample` should use a `< 0` check instead of a `<= 0` check since this allows the whole decrement and compare to 0 operation which is on the hot path of every allocation to be serviced by a single RMW like `sub [interval_to_next_sample_], rax` followed by a comparison using the carry flag. IIRC that saved 3 instructions on the hot path though the net improvement is less than this would suggest as an RMW is more expensive than your average instruction (though less expensive than the replacement sequence).

To make that change correct, I had to decrement the initial value of `interval_to_next_sample_` by 1 which is why the  +1 disappeared.

Note that the change is from `next + 1` to `next` not the other way around (your comment suggests the other way around).

Reply:
If we go with this logic, then I think you should also change the inequality `while (interval_to_next_sample_ <= 0)` to `< 0`. (Otherwise, e.g. when you sample allocations of size 1 with sampling interval 1, a sampled allocation will always be followed by an unsampled one).

(Yes, of course this doesn't matter for any practical purpose).

Reply:
Good call! In force push https://github.com/scylladb/seastar/commit/a7d7f691117f8971d240937bb789cd3e9d8c87ed I have updated that loop to use `< 0`.

cc @StephanDollberg